keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,576 @@
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+The following common error codes are proposed for volume health:
+* VolumeNotFound
+* VolumeUnmounted
+* OutOfCapacity
+* RWIOError
+* DiskFailure
+* DiskRemoved
+* ConfigError
+* NetworkError
+* DiskDegrading
+* FilesystemCorruption
+* NodeFailure
+
+Add `GET_VOLUME` and `GET_VOLUME_HEALTH` controller capabilities.  If a driver supports `GET_VOLUME` capability, it MUST implement the `GetVolume` RPC.  If a driver supports `GET_VOLUME_HEALTH`, it MUST supports the `health` field in the `status` field in `GetVolumeResponse`.
+
+```
+// Specifies a capability of the controller service.
+message ControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+      CREATE_DELETE_VOLUME = 1;
+      PUBLISH_UNPUBLISH_VOLUME = 2;
+      LIST_VOLUMES = 3;
+      GET_CAPACITY = 4;
+      // Currently the only way to consume a snapshot is to create
+      // a volume from it. Therefore plugins supporting
+      // CREATE_DELETE_SNAPSHOT MUST support creating volume from
+      // snapshot.
+      CREATE_DELETE_SNAPSHOT = 5;
+      LIST_SNAPSHOTS = 6;
+
+      // Plugins supporting volume cloning at the storage level MAY
+      // report this capability. The source volume MUST be managed by
+      // the same plugin. Not all volume sources and parameters
+      // combinations MAY work.
+      CLONE_VOLUME = 7;
+
+      // Indicates the SP supports ControllerPublishVolume.readonly
+      // field.
+      PUBLISH_READONLY = 8;
+
+      // See VolumeExpansion for details.
+      EXPAND_VOLUME = 9;
+
+      // Indicates the SP supports the
+      // ListVolumesResponse.entry.published_nodes field
+      LIST_VOLUMES_PUBLISHED_NODES = 10;
+
+      // Indicates the SP supports the GetVolume RPC
+      GET_VOLUME = 11;
+
+      // Indicates the SP supports the
+      // GetVolumeResponse.volume_health field
+      GET_VOLUME_HEALTH = 12;
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+#### Add Node Volume Health Function
+
+Node Volume Health checks if a volume is still mounted and usable. To check whether a volume is usable, the CSI driver shall check if filesystem is corrupted, whether there are bad blocks, etc. in this RPC.
+
+Instead of adding a new RPC, we can leverage the existing NodeGetVolumeStats RPC.
+
+```
+rpc NodeGetVolumeStats (NodeGetVolumeStatsRequest)
+    returns (NodeGetVolumeStatsResponse) {}
+```
+
+In the NodeGetVolumeStatsRequest, there are `volume_id`, `volume_path`, and `staging_target_path`. CSI driver can figure out whether a volume is still mounted based on these parameters.
+
+In NodeGetVolumeStatsResponse, there is already a `VolumeUsage` which includes information (available, used, and total) of the file system on the volume.
```

Comment:
Have you though how `VolumeUsage` will be exposed/consumed within cluster? I think this is interesting topic that also applies to container runtime and network. It's ok if not.

There was one idea on my mind to introduce USE (utilization, saturation, error) metrics from interfaces like (CRI, CNI, CSI) and expose them in cluster using [kubernetes metrics framework](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/component-base/metrics) (wrapper for prometheus).
I think this case shows that there is a need to expand scope of monitoring in k8s (currently limited to autoscaling)
/cc @brancz @davidashpole
I've been drafting a proposal on this and I would be interested in discussing it more in SIG-instrumentation.